### PR TITLE
Add documentation for cross-join elimination

### PR DESCRIPTION
**Pull request by mkleen**
## Checklist

 - [x] Added an entry in the latest `docs/appendices/release-notes/<x.y.0>.rst` for user facing changes
 - [x] Updated documentation & `sql_features` table for user facing changes
 - [x] Touched code is covered by tests
 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed
 - [x] This does not contain breaking changes, or if it does:
    - It is released within a major release
    - It is recorded in the latest `docs/appendices/release-notes/<x.y.0>.rst`
    - It was marked as deprecated in an earlier release if possible
    - You've thought about the consequences and other components are adapted
      (E.g. AdminUI)

### CODE REVIEW

File: docs/concepts/joins.rst [hunk 1]
```diff
@@ -350,6 +350,32 @@ optimized.*
    Complex queries are broken down into subqueries that are run on their shards
    before joining.
 
+.. _join-optim-eliminate-cross-joins:
+
+Cross join elimination
+----------------------------
```
Comment (mfussenegger) [inline docs/concepts/joins.rst:356]:
```suggestion
Cross join elimination
----------------------
```

File: docs/concepts/joins.rst [hunk 2]
```diff
@@ -350,6 +350,32 @@ optimized.*
...
+
+The optimizer will try to eliminate cross joins in the query plan by changing
+the join-order. When reordering, it will try to preserve the original join order
+as much as possible. If a cross join cannot be eliminated, the original join
+order will be maintained. It will try to eliminate any cross join
+it can e.g.:
```
Comment (mfussenegger) [inline docs/concepts/joins.rst:362]: Maybe add a sentence or two that defines or explains what cross join elimination is, or what it does. 

E.g. something like "Cross join elimination replaces a CROSS JOIN with a INNER JOIN if query conditions used in the WHERE clause or other join conditions allow for it. An example: ..."

The whole paragraph kinda assumes that people are already familiar with it.

File: docs/concepts/joins.rst [hunk 3]
```diff
@@ -350,6 +350,32 @@ optimized.*
...
+    INNER JOIN t3
+    ON t3.z = t1.x AND t3.z = t2.y
+
+The cross join elimination will change the order of the query from t1, t2, t3
+to t2, t1, t3 so that each join has a join condition and the cross-join can be
+eliminated. This optimizer rule can be disabled with the session settings::
+
+    SET optimizer_eliminate_cross_join = false
+
+Note that this setting is experimental, and may change in the future.
```
Comment (mfussenegger) [inline docs/concepts/joins.rst:377]: I think the setting should be documented in docs/config/session.rst and this should link to it / vice versa

File: docs/concepts/joins.rst (hunk 2 above)
Comment (mkleen) [reply docs/concepts/joins.rst:362]: 👍

File: docs/concepts/joins.rst (hunk 3 above)
Comment (mkleen) [reply docs/concepts/joins.rst]: 👍

**Comment by mkleen** [comment]
This is ready for another review.

File: docs/config/session.rst [hunk 4]
```diff
@@ -212,6 +212,13 @@ releases.
    The optimizer setting is for advanced use only and can significantly impact
    the performance behavior of the queries.
 
+**optimizer_eliminate_cross_join**
```
Comment (amotl) [inline docs/config/session.rst:215]: Vale thinks it found a mistake here. Is there a canonical way to suppress it?

**Review by amotl — APPROVED** [review]

File: docs/config/session.rst (hunk 4 above)
Comment (BaurzhanSakhariev) [reply docs/config/session.rst:215]: Maybe wrapping with `` would help?

Comment (matriv) [reply docs/config/session.rst:215]: If this doesn't help there is
```
..vale off
<context-to-suppress>
..vale on
```

**Pull request merged**